src/grisp_hw.erl

```diff
@@ -1,20 +1,97 @@
+-export([reset_eeprom_crc/0]).
+
+% testing
+-export([crc16/1]).
 
--on_load(init/0).
+% Callbacks
+-export([load_nif/0]).
+-on_load(load_nif/0).
 
 %--- API -----------------------------------------------------------------------
 
 % @doc Return the platform name as an atom.
 -spec platform() -> atom().
 platform() -> erlang:nif_error("NIF library not loaded").
 
+
+% @doc Read GRiSP meta data from EEPROM
+%
+% === Example ===
+% ```
+% 2> grisp_eeprom:read().
+% [{grisp_version, "2"},
+%  {grisp_serial, 1001},
+%  {grisp_pcb_version, "1.2"},
+%  {grisp_pcb_variant, 1},
+%  {grisp_batch, 1},
+%  {grisp_prod_date, {{2021,10,1}, {0,0,0}}}]
+% '''
+-spec read_eeprom() -> [].
+read_eeprom() ->
+    Bus = grisp_ni2c:open(?GRISP_EEPROM_BUS),
+    <<_SigVersion:8,
+      _Dummy1:3/binary,     %% unused
+      Serial:4/little-unit:8,
+      BatchNr:2/little-unit:8,
+      ProdYear:2/little-unit:8,
+      ProdMonth:8,
+      ProdDay:8,
+      VersMajor:1/binary,   %% contains GRiSP version and PCB Major
+      VersMinor:8,          %% actually the PCB minor
+      Variant:8,
+      _Mac:6/binary,
+      _Dummy2:1/binary,     %% unused
+      Crc:2/little-unit:8,
+      _Dummy3:2/binary>> = Data = grisp_ni2c:read(Bus, ?GRISP_EEPROM_ADR,
+                                                  0, ?GRISP_EEPROM_DATA_SIZE),
+    <<GrispVersion:4, PcbMajor:4>> = VersMajor,
+    MetaData = [{grisp_version, lists:flatten(io_lib:format("~p", [GrispVersion]))},
+                {grisp_serial, Serial},
+                {grisp_pcb_version, lists:flatten(io_lib:format("~p.~p", [PcbMajor, VersMinor]))},
+                {grisp_pcb_variant, Variant},
+                {grisp_batch, BatchNr},
+                {grisp_prod_date, {{ProdYear,ProdMonth,ProdDay},{0,0,0}}}],
+    <<DataToBeVerified:24/binary, _/binary>> = Data,
+    case (crc16(DataToBeVerified) =:= Crc) of
+        true    -> {ok, MetaData};
+        false   -> {invalid_crc, MetaData}
+    end.
+
+
+% @doc Fixes CRC bytes for pre-production boards
+-spec reset_eeprom_crc() -> ok.
+reset_eeprom_crc() ->
+    Bus = grisp_ni2c:open(?GRISP_EEPROM_BUS),
+    <<DataToBeVerified:24/binary, _/binary>> = grisp_ni2c:read(Bus, ?GRISP_EEPROM_ADR,
+                                                               0, ?GRISP_EEPROM_DATA_SIZE),
+    CrcData = <<(crc16(DataToBeVerified)):2/little-unit:8>>,
+    grisp_ni2c:transfer(Bus, [
+        {write, ?GRISP_EEPROM_ADR, 0, <<24:8>>},
+        {write, ?GRISP_EEPROM_ADR, 16#4000, CrcData}]),
+    ok.
+
+
 %--- Callbacks -----------------------------------------------------------------
 
-init() -> ok = erlang:load_nif(atom_to_list(?MODULE), 0).
+load_nif() -> ok = erlang:load_nif(atom_to_list(?MODULE), 0).
+
+
+% @private
+crc16(Cp) ->
+    crc16(0, Cp).
+
+% @private
+crc16(Crc, <<>>) ->
+    Crc;
+crc16(Crc, <<Cp:8, CpTail/binary>>) ->
+    crc16((Crc bsr 8) bxor lists:nth(((Crc bxor Cp) band 16#ff) + 1, ?misc_crc16_tab), CpTail).
```

Comment:
I think a tuple accessed with `element/2` would be much faster here.

Reply:
I can change that even though speed here has no real relevance.